### PR TITLE
This PR enhances the development branch of the PC-Transformer with the following updates:

### PR DESCRIPTION
**Pull request by hailer-MIT**
/request-review @Nardos24 @quinbez

## Summary
This PR integrates enhanced predictive coding dynamics, updated tokenization, and optimized data loading.

### 20bdfe7 - Enhance new_attn branch with Sum-based energy and Temporal Energy Integration

 - **Summary:**
             --> Transitioned from mean-based to sum-based error aggregation globally across all PC layers to align with the Total Energy Minimization framework.
             --> Implemented stateful temporal tracking by capturing discrete energy values for each inference time-step.
- **Detail:**
             --> Modified finalize_step to calculate the total error across the entire tensor volume, providing a more comprehensive energy landscape than element-wise averaging.
             --> Enabled high-resolution debugging by replacing scalar energy summation with a sequential history of errors across all $T$ iterations. This enables detailed analysis of the model's convergence behavior during the iterative inference process.

### 16c978b - BERT style tokens replaced and Vocab_size enhanced
- **Summary:** 
                      -->  BERT-style tokens replaced by GPT-style token standardization
                      --> and vocab size is enhanced to 11707  
- **Detail:** Manually resolved conflicts in `config.py` and `prepare_tokens.py` to preserve vocab_size=11707 while adopting team's tokenizer improvements.

### 2f6694e - Updated BPE training to study all three dataset splits (Train, Valid, Test)
- **Summary:** Modified BPE tokenizer training to use train/validation/test splits
- **Detail:** Updated `prepare_tokens.py` to train tokenizer on combined dataset splits, improving vocabulary coverage.

### 0a400f8 - Implemented Steady-State energy reporting in pc_layer.py
- **Summary:** enhance the energy reporting to report from final inference step only.
- **Detail:** Modified `pc_layer.py` to report energy only after all T iterations, excluding transient intermediate states.

### af37ccc - feat: vocab_size excludes special token count and Applied td_err delta_x to include the missing top down error in for the latent state update.
- **Summary:**
       --> Integrated new_attn branch with updated vocab size 
       -->and TD error handling implemented
- **Detail:** 
        ---> vocab_size update to 11707,  
        --> And added delta_x = delta_x - td error if td error doesn't exist at the step attn function of  pc_utils.py

### 000f80a - Optimize data loading with sequence padding
- **Summary:** Added sequence padding for last batch to reserve the last batch from being dropped and disabled drop_last to prevent data loss
- **Detail:** Modified `dataloader.py` to pad sequences to block_size and disabled `drop_last` to ensure all data samples are used.

### Technical Note: Energy Scale Discrepancy (Sum vs. Mean)

I wanted to provide a technical clarification on the significant difference in reported Energy values between our previous implementation and the current dev_enhanced branch.

The Discrepancy:

  - Original Approach: Used Mean-based aggregation at the tensor level and Sum-based aggregation across time-steps. (Resulted in Energy approximately 1.0)
 - Current Approach: Uses Sum-based aggregation at the tensor level and reports only the Last Time-Step. (Resulted in Energy approximately 50k).
 
Mathematical Rationale: 

The large jump in value is a direct result of the Tensor Volume. In our current architecture, a layer's activity matrix contains hundreds of thousands of individual elements.

 - Summing (Current): When we switch from mean() to sum() at the finalize_step, we are accumulating every single one of those hundreds of thousands of errors into a single scalar. This naturally results in a very large number (50k+).
 - Averaging (Previous): Averaging essentially compresses that entire matrix of errors back down to the scale of a single element, which is why the previous numbers stayed near 1.0.
 
Conclusion: 

Reporting only the final time-step in the current branch does not 'fix' the large numbers, because that single time-step is still a Sum of the entire tensor. We should not expect the current Energy to ever reduce down to the levels of the previous Approach, as they are measured on completely different mathematical scales. Our focus should be on the downward trend of the 50k value, rather than the absolute number itself.

### CODE REVIEW

**Review by Nardos24 — APPROVED** [review]

File: predictive_coding/pc_layer.py
Comment (quinbez) [inline predictive_coding/pc_layer.py:1]: Since `get_energy()` only ever reads the last element of `_energy_list`, the list itself is never meaningfully used as a list, nothing iterates over it or accesses any index other than -1. Every forward pass it grows to T entries then gets wiped, which means all the intermediate appends are wasted allocations. The intent of reporting only the final T step is correct, but this could be achieved more simply by just keeping a single scalar that gets overwritten on each step rather than accumulated into a list. Same result, less overhead.

File: data_preparation/config.py [hunk 1]
```diff
@@ -11,5 +11,13 @@
...
-special_tokens = ["[UNK]", "[CLS]", "[SEP]", "[PAD]", "[MASK]"]
\ No newline at end of file
+
+vocab_size = 11707 # Optimized subword-centric vocabulary size after GPT2
+
+# [MASK] exists only for BERT masked language modeling.
+# special_tokens = ["[UNK]", "[CLS]", "[SEP]", "[PAD]", "[MASK]"]
+special_tokens = ["[UNK]", "[BOS]", "[EOS]", "[PAD]"]
+
+# Training parameters
```
Comment (quinbez) [inline data_preparation/config.py]: Hardcoding `batch_size` and `max_len` here is unnecessary since `training.py` and `GPTConfig` already manage them through `best_hyperparameters.txt`.

File: data_preparation/config.py [hunk 2]
```diff
@@ -11,5 +11,13 @@
...
 
 # Tokenizer parameters
-vocab_size = 1024
-special_tokens = ["[UNK]", "[CLS]", "[SEP]", "[PAD]", "[MASK]"]
\ No newline at end of file
+
+vocab_size = 11707 # Optimized subword-centric vocabulary size after GPT2
+
+# [MASK] exists only for BERT masked language modeling.
+# special_tokens = ["[UNK]", "[CLS]", "[SEP]", "[PAD]", "[MASK]"]
```
Comment (quinbez) [inline data_preparation/config.py]: Now that the switch to GPT-style tokens is done, the old commented-out BERT special tokens line can be removed to keep the config clean.

File: data_preparation/dataloader.py [hunk 3]
```diff
@@ -33,7 +33,7 @@ def get_loaders(batch_size: int, block_size: int, distributed: bool = False, str
         batch_size= batch_size, 
         sampler=train_sampler,
         shuffle=(train_sampler is None), 
-        drop_last=True
+        drop_last=False
```
Comment (quinbez) [inline data_preparation/dataloader.py:36]: The cross entropy in `training.py` uses `ignore_index=0` to mask padding, but the pad token ID in the dataset is 3. This means pad tokens are not being ignored during loss computation and will corrupt training. `ignore_index` should be updated to 3, because currently `drop_last=False` with padding means the model will train on padded sequences.

File: data_preparation/dataset.py [hunk 4]
```diff
@@ -19,15 +21,17 @@ def __init__(self, file_path, block_size, stride=None):
...
-                device=tokens.device,
-            )
-        else:
-            # `unfold` keeps overlapping windows as a view instead of copying the corpus.
-            self.sequences = tokens.unfold(0, window_size, self.stride)
+
+        # Calculate how much padding we need to avoid throwing away data
+        chunk_size = block_size + 1
+        remainder = len(tokens) % chunk_size
+
```
Comment (quinbez) [inline data_preparation/dataset.py]: Sliding window is replaced without acknowledgment. The original `tokens.unfold(0, window_size, stride)` created overlapping windows which was a deliberate design choice for richer sequence coverage. Replacing it with `tokens.view(-1, chunk_size)` creates non-overlapping fixed chunks. This is a significant behavioral change that isn't mentioned in the PR description. Also the original guard for when `len(tokens) < window_size` is now gone, so very small token files will fail silently or produce wrong shapes.

File: predictive_coding/pc_layer.py
Comment (hailer-MIT) [reply predictive_coding/pc_layer.py:1]: Good catch, the list was leftover scaffolding from when I was debugging per-step convergence. Replaced with a single overwritten scalar (_energy_scalar); same semantics, no wasted allocations.

File: data_preparation/config.py (hunk 1 above)
Comment (hailer-MIT) [reply data_preparation/config.py]: Agreed, those were convenience variables from early prototyping. Since training.py sources all hyperparameters from best_hyperparameters.txt / GPTConfig, these lines are dead code that could mislead readers. Removed.

File: data_preparation/config.py (hunk 2 above)
Comment (hailer-MIT) [reply data_preparation/config.py]: The commented-out BERT line was left as a migration reference. Since the switch is complete and documented in the PR, it's just noise — removed.

File: data_preparation/dataloader.py (hunk 3 above)
Comment (hailer-MIT) [reply data_preparation/dataloader.py:36]: Critical catch, ignore_index was pointing at [UNK] (id=0) instead of [PAD] (id=3). This meant padded positions contributed to the loss while real [UNK] tokens were silently skipped. Fixed ignore_index to 3 to match pad_token_id in dataset.py.

File: data_preparation/dataset.py (hunk 4 above)
Comment (hailer-MIT) [reply data_preparation/dataset.py]: You are completely right on both points. I replaced tokens.unfold(0, window_size, stride) with tokens.view(-1, chunk_size) as part of adding padding support, but did not acknowledge this in the PR description. This was a significant oversight, the overlapping sliding window was a deliberate design for richer sequence coverage, and my replacement silently ignored the stride parameter entirely. 

I have now restored tokens.unfold() as the primary sequence generator. Tail padding is applied using a minimal calculation that pads just enough tokens to turn any leftover tail into one final complete window, so no tokens are discarded while the overlapping window design is fully preserved. The small-file guard has also been restored.

Finally, Thank you @quinbez for these detailed and precise feedbacks.

File: data_preparation/prepare_tokens.py [hunk 5]
```diff
@@ -25,7 +25,7 @@ def build_tokenizer():
 
     trainer = BpeTrainer(
         special_tokens = special_tokens,
-        vocab_size = vocab_size
+        vocab_size = vocab_size + len(special_tokens)
```
Comment (quinbez) [inline data_preparation/prepare_tokens.py]: While testing the changes, I encountered an out-of-bounds error from `F.one_hot` caused by token indices exceeding the declared vocab size. The issue is that `config.py` sets `vocab_size = 11707`, while `prepare_tokens.py` trains the tokenizer with `vocab_size + len(special_tokens)` = 11711. As a result, the tokenizer can produce indices larger than the model expects. 

I would recommend removing the `+ len(special_tokens)` addition from `BpeTrainer` and instead updating `config.py` to include the special tokens in the total vocab size.

Comment (hailer-MIT) [reply data_preparation/prepare_tokens.py]: fixed, thanks.

Comment (quinbez) [reply data_preparation/prepare_tokens.py]: The fix needs to be applied in both files. Updating `config.py` alone will not solve the issue because `prepare_tokens.py` still adds `len(special_tokens)` inside `BpeTrainer`, causing the tokenizer vocab size to become 11715 instead of 11711. The extra addition should be removed so both files use the same vocabulary size.

Comment (hailer-MIT) [reply data_preparation/prepare_tokens.py]: sorry for the mistake. you already mentioned it.

**Review by quinbez — APPROVED** [review]
